server/internal/cryptoworker/crypto_worker.go

```diff
@@ -103,10 +103,10 @@ func (w *Worker) doUnwrap(pkt *packet.Packet) error {
 
 		w.log.Debugf("Packet: %v (Unwrap took: %v)", pkt.ID, unwrapAt.Sub(startAt))
 
+		lastErr = err
```

Comment:
How about if we just return error for the below conditions 

```go
		if err = pkt.Set(payload, cmds); err != nil {
			return fmt.Errorf("crypto: Failed to set payload: %w", err)
		}

		// Check for replayed packets.
		if k.IsReplay(tag) {
			// The packet decrypted successfully, the MAC was valid, and the
			// tag was seen before, therefore drop the packet as a replay.
			instrument.PacketsReplayed()
			return errors.New("crypto: Packet is a replay")
		}

Reply:
that wouldn't do the same thing because we only want to return the error after the for loop is done.